CARDS-1321: As a questionnaire designer, I can specify default values for questions with list or list+input display mode

veronikaslc opened the pull request.

Commit:
CARDS-1321: As a questionnaire designer, I can specify default values for questions with list or list+input display mode

veronikaslc commented:
> the default option does not persist after saving the form

I am not sure it shoud. The issue description only said it has to be displayed as selected in UI. If we want to set the default value for certain answers we have to set them at the backend at the time of form creation. @marta- WDYT

> multiple defaults are not supported

Also here I am not sure it should support multiple defaults, I intentionally implemented only support for a single option.

marta- commented:
> > the default option does not persist after saving the form
> 
> I am not sure it shoud. The issue description only said it has to be displayed as selected in UI. If we want to set the default value for certain answers we have to set them at the backend at the time of form creation. @marta- WDYT

From the user perspective, it make no sense to display an option as selected but not save it. I will clarify the specs, as it seems I omitted to mention that.

Regarding the implementation: nothing needs to be done in the backend, just work with the `initialSelection` in `MultipleChoice`.
> 
> > multiple defaults are not supported
> 
> Also here I am not sure it should support multiple defaults, I intentionally implemented only support for a single option.

If the Question supports multiple answers, it should support multiple defaults. To have an easy way to deal with the case where the question spec is flawed, i.e. the number of default values is greater than `maxAnswers` (and `maxAnswers` > 0), only take into account the first `maxAnswers` default values.

Commit:
CARDS-1321: As a questionnaire designer, I can specify default values for questions with list or list+input display mode
Addressed codereview comments: allow to save defaults and allow more than onde default

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/MultipleChoice.jsx:
```diff
@@ -78,6 +77,15 @@ function MultipleChoice(props) {
+    if (maxAnswers > 0 && initialSelection.length > maxAnswers) {
+      initialSelection = initialSelection.slice(0, maxAnswers);
+    }
+  }
```
This block + line 64 could be replaced with the following (not tested):

```suggestion
  if (initialSelection.length == 0) {
    initialSelection = defaults.filter(item => item[IS_DEFAULT_ANSWER_POS])
      // If there are more default values than the specified maxAnswers, only take into account the first maxAnswers default values.
      .slice(0, maxAnswers || defaults.length)
      .map(item => [item[LABEL_POS], item[VALUE_POS]]);
  }
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/MultipleChoice.jsx:
```diff
@@ -527,7 +535,8 @@ function generateDefaultOptions(defaults, selection, disabled, isRadio, onClick,
         onDelete={onDelete}
-        isDefault={childData[IS_DEFAULT_POS]}
+        isDefaultOption={childData[IS_DEFAULT_OPTION_POS]}
+        isDefaultAnswer={selection.length == 0 && childData[IS_DEFAULT_ANSWER_POS]}
```
The values in `defaults[i][IS_DEFAULT_ANSWER_POS]` don't account for the `maxAnswers` limitation, so this flag alone is not reliable for deciding whether an option is a default value. 

Regardless. the prop `isDefaultAnswer` seems unnecessary. This component doesn't need to know specifically if this option is a default value or not, only if it's checked, which should already be captured by `selection`.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/MultipleChoice.jsx:
```diff
@@ -539,27 +548,28 @@ var StyledResponseChild = withStyles(QuestionnaireStyle)(ResponseChild);
                   isRadio ?
                   (
                     <Radio
+                      checked={checked || isDefaultAnswer}
```
Wouldn't the `checked` flag be true anyway if this is a default answer, because we're adding this option to `initialSelection`, which is used to initialize `selection`?

marta- commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/AnswerOptions.jsx:
```diff
@@ -198,21 +199,22 @@ let AnswerOptions = (props) => {
           <ComposedIcon
               size="large"
-              MainIcon={NotesIcon}
+              MainIcon={MoreVertIcon}
```
I don't think the `ExtraIcon` displaying `Add` or `Edit` fits anymore. Moreover, this `ExtraIcon` only reflects the changes in the description, not in the `isDefault` setting.

I would use no `ExtraIcon` if there's no description and `isDefault` is `false`, and `ShortText` if there's a description or `isDefault` is changed to `true`. Warning: we may change the the icon a couple if times more until it looks right.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/AnswerOptions.jsx:
```diff
@@ -409,6 +418,18 @@ let AnswerOptions = (props) => {
             onChange={setDescription}
           />
         }
+        <Tooltip title="" className={classes.isDefaultSwitch}>
+        <FormControlLabel
+          label="Is default:"
+          labelPlacement="start"
+          control={
+            <Switch
+              checked={!!isDefault}
+              onChange={(event) => setIsDefault(event.target.checked)}
+              color="primary"
+              />
+          }
+        /></Tooltip>
```
Can't we resue the [BooleanInput](https://github.com/data-team-uhn/cards/blob/dev/modules/data-entry/src/main/frontend/src/questionnaireEditor/BooleanInput.jsx) here?

Also, the [specs](https://phenotips.atlassian.net/browse/CARDS-1321) mentioned changing the title from `Description for ...` to `More options for...`, (since we have more than just description now) and add a label `Description` between the title and the markdown editor.

veronikaslc replied:
> Can't we resue the BooleanInput here
Unfortunatelly not since it's used in a popover dynamically filled with data. BooleanInput is for static hidden input generations.

Commit:
CARDS-1321: As a questionnaire designer, I can specify default values for questions with list or list+input display mode
Addressed codereview comments

sashaandjic reviewed (approved):
tested

sdumitriu committed:
CARDS-1321: As a questionnaire designer, I can specify default values for questions with list or list+input display mode
Bugfix: after unselecting the default options from a list, when re-editing an existing form the default options should not appear re-selected

sdumitriu reviewed (approved):
I don't quite like how the more details icon looks when there are setting specified, but we can look for a better one later.

A minor bug: in the questionnaire editor, editing an answer option to say that is default doesn't immediately update the icon, only after closing and reopening the question editor does it show that it has custom settings.

sashaandjic commented:
> A minor bug: in the questionnaire editor, editing an answer option to say that is default doesn't immediately update the icon, only after closing and reopening the question editor does it show that it has custom settings.

Oh I missed that. Can we fix it?

marta- commented:
> > A minor bug: in the questionnaire editor, editing an answer option to say that is default doesn't immediately update the icon, only after closing and reopening the question editor does it show that it has custom settings.
> 
> Oh I missed that. Can we fix it?

Agreed, should be fixed before merge

Commit:
CARDS-1321: As a questionnaire designer, I can specify default values for questions with list or list+input display mode
Fixed description icon change on isDefault property update

Commit:
CARDS-1321: As a questionnaire designer, I can specify default values for questions with list or list+input display mode
Fixed question edit icon whitescreen on click

marta- commented on modules/commons/src/main/frontend/src/components/ComposedIcon.jsx:
```diff
@@ -63,20 +63,19 @@ const ComposedIcon = ({
 
 ComposedIcon.propTypes = {
   MainIcon: PropTypes.elementType.isRequired,
-  ExtraIcon: PropTypes.elementType.isRequired,
```
This propType should not have been completely removed, only `isRequired` needs to be discarded.

sashaandjic reviewed (approved):
tested

marta- commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/AnswerOptions.jsx:
```diff
@@ -39,8 +39,8 @@ import QuestionComponentManager from "./QuestionComponentManager";
 import MarkdownText from "./MarkdownText";
 import CloseIcon from '@material-ui/icons/Close';
 import AddCircleIcon from '@material-ui/icons/AddCircle';
```
`AddCircleIcon` is no longer used, remove.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/AnswerOptions.jsx:
```diff
@@ -39,8 +39,8 @@ import QuestionComponentManager from "./QuestionComponentManager";
 import AddCircleIcon from '@material-ui/icons/AddCircle';
-import EditIcon from '@material-ui/icons/Edit';
-import NotesIcon from '@material-ui/icons/Notes';
+import ShortTextIcon from '@material-ui/icons/ShortText';
```
Let's go back to using `EditIcon` instead of `ShortTextIcon` to mark when options have been modified.

marta- replied:
After discussions, it was decided to not use a composite icon for the "More options" after all. Instead, we'll make the More options" icon color dark grey (`theme.palette.text.primary`) when there's a description or the option is marked as default.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/AnswerOptions.jsx:
```diff
@@ -198,22 +199,23 @@ let AnswerOptions = (props) => {
 
   let generateDescriptionIcon = (item, index, isSpecialOptn) => {
```
Since this icon is no longer only about Description, update the function name to `generateOptionsIcon`

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/MultipleChoice.jsx:
```diff
@@ -539,19 +545,20 @@ var StyledResponseChild = withStyles(QuestionnaireStyle)(ResponseChild);
                   isRadio ?
                   (
                     <Radio
+                      checked={checked}
```
Why is this needed? Doesn't the `RadioGoup` take care of the `checked` status for its Radio buttons?

marta- reviewed (dismissed):
Let's switch the places of the AnswerOption actions, so the "More actions" button is last. With that plus the modification of the "More options" look (as mentioned in an inline comment), please see the expected outcome in this mockup:

![image (48)](https://user-images.githubusercontent.com/651980/133344066-ac40af8f-99d9-4b99-b702-50fd77e8de52.png)

Commit:
CARDS-1321: As a questionnaire designer, I can specify default values for questions with list or list+input display mode
Addressed codereview comments

sashaandjic commented:
Looks good:

![image](https://user-images.githubusercontent.com/23243244/133383144-7d67f880-eeff-4840-9a76-0810d4847cc1.png)

Commit:
CARDS-1321: As a questionnaire designer, I can specify default values for questions with list or list+input display mode
Removed checked for Radio

marta- committed:
[misc] Moved answer options styling from QuestionnaireStyle to AnswerOptions since it is not used in any other components

marta- committed:
CARDS-1321: As a questionnaire designer, I can specify default values for questions with list or list+input display mode
* Used a checkbox in front of the option label for `isDefault`, for a more "what you see is what you get" user experience
* Ensured `isDefault` is saved and loaded as boolean instead of string
* Made the special options look more like the other options
* Simplified styling

sashaandjic reviewed (approved):
tested
perfect: I can see defaults (checkbox) and edited options (pencil) at the same time 
![image](https://user-images.githubusercontent.com/23243244/133543058-26ea771b-97fa-4ae6-ba50-0f487e6dfdbd.png)

marta- commented:
The latest changes include storing `isDefault` as boolean instead of string and refactoring the description popover. Everything needs to be retested (including if the default answers appear correctly on the form and  if the description works correctly in the wizard).

marta- committed:
[misc] Simplified option description popover styling; improved icon button title and handler function names.

sashaandjic commented:
It looks and feels good. I will do more testing later today.

sashaandjic reviewed (approved):
tested
